### DISCUSSION TITLE
How do i configure imgui in clion?

### DISCUSSION THREAD

**Discussion by bret-walda** [ocornut/imgui#4077] [New Users Build/Link/Run issues ONLY!]
i have copied everything from imgui master that is required in project but when i run it gui window doesn't appear and it only says code runs successfully i am using dx11 and win32 in clion but how do successfully configure imgui in clion please help i have looked everywhere and didn't find a solution can anyone help any help is greatly appreciated Thanks.

**Comment by rokups** [comment]
Put [CMakeLists.txt](https://gist.github.com/rokups/f771217b2d530d170db5cb1e08e9a8f4) in imgui folder and open it in CLion.

**Reply by bret-walda** [reply]
so i put CMakelists.txt in the imgui folder i downloaded from here or created in clion with all cpp and header files i am sorry if i sound a little stupid i am new to imgui and having problems but will try this and tell you back if it works or not.

**Reply by bret-walda** [reply]
I have added all header files and cpp files by putting them directly into the project and adding them into add_executable() for the sake of just running my first program successfully but i just keep getting this error 

'D3D11_VIEWPORT_AND_SCISSORRECT_OBJECT_COUNT_PER_PIPELINE' was not declared in this scope
         D3D11_RECT                  ScissorRects[D3D11_VIEWPORT_AND_SCISSORRECT_OBJECT_COUNT_PER_PIPELINE];

i am exhausted right now of how much googling i have done to resolve this error but nothing please help.

**Reply by rokups** [reply]
Must be missing includes. Hard to tell without knowing full picture. However all examples are there, please take a look at them.

---

If you would like to build your imgui application using cmake try this:

Suppose you have this directory tree:
```
|_ imgui                   (git clone https://github.com/ocornut/imgui)
|   |_ imgui.cpp           (and others)
|   |_ CMakeLists.txt      (https://gist.github.com/rokups/f771217b2d530d170db5cb1e08e9a8f4)
|_ main.cpp                (your application, you can copy one from imgui/examples/example_win32_directx11/main.cpp)
|_ CMakeLists.txt          (your cmake script)
```
In your top level `CMakeLists.txt` add this:
```cmake
cmake_minimum_required (VERSION 3.2)
project (my-project)

add_subdirectory(imgui)
add_executable(my-program main.cpp)
target_link_libraries(my-program PUBLIC imgui imgui-win32 imgui-dx11)
```

**Reply by bret-walda** [reply]
i have given up no matter what i do i cant get it to work i also tried this but still nothing maybe as a last try you can teamviewer into my pc to see what am i doing wrong. only if you have time and want to otherwise peace be upon you and goodbye. Thanks for helping me.

**Comment by petervdpas** [comment]
cmake_minimum_required(VERSION 3.24)
project(MyImGui)

set(CMAKE_CXX_STANDARD 20)

add_executable(MyImGui main.cpp imgui.cpp imgui_demo.cpp imgui_draw.cpp imgui_impl_glfw.cpp imgui_impl_opengl3.cpp imgui_tables.cpp imgui_widgets.cpp)

find_package(GLEW REQUIRED)
target_link_libraries(MyImGui PRIVATE GLEW::GLEW)

find_package(glfw3 CONFIG REQUIRED)
target_link_libraries(MyImGui PRIVATE glfw)

find_package(imgui CONFIG REQUIRED)
target_link_libraries(MyImGui PRIVATE imgui::imgui)

#Hope this helps

**Reply by ori1202** [reply]
does it work on c++11?

**Comment by RedBloodCat** [comment]
Under step 3, where can i find the <optional build paramiters> and what they are/do?

Line im referencing (17:1):
#    cmake <optional build parameters> ../imgui

**Comment by TheZitroX** [comment]
Did it work?

**Reply by bret-walda** [reply]
i am not using clion anymore but yes i have made imgui work now in vim.

**Reply by TheZitroX** [reply]
Was clion on Mac?
I want to develop for fun on my MacBook, but somehow it's weird that cpp std files doesn't even get found by Cmake...

**Reply by bret-walda** [reply]
yes it is on macOS. by std files if you mean standard library then standard library doesn't have anything to do with cmake make sure gcc or clang are installed on your system but if you mean your source files are not being recognized then share your cmake text file and i will try to help.

**Reply by TheZitroX** [reply]
It has to do with the std libs yes.
I describe it after lunch ✌️

**Reply by bret-walda** [reply]
ok no problem.

**Reply by TheZitroX** [reply]
I installed the build chain and found most of the paths. But still it has issues with base libs like memory headder.

I think it's up to me to find out a way of doing things on MacBooks.

It should work on m2 chips right?
Otherwise I go back to my windows pc (wich is actually way faster, but I love to sit on the couch when doing random stuff)

**Reply by bret-walda** [reply]
Yes it should but if you need any help just reply to me here.

**Reply by TheZitroX** [reply]
I did it @bret-walda !! :D
Took a while and i had to install the newest cmake gcc and libs
weird that it had problems and many missing libs...

Now im facing issues with resizing. Looks like a mac problem vs windows

**Comment by romkor-lab-it** [comment]
Guys, can somebody please explain from the start? Because I have wasted a week with completely zero success, and I don't know where to find solution. Can somebody just help me with that?

**Reply by bret-walda** [reply]
Put your cmake file here and I will try to help the best I can.

**Reply by romkor-lab-it** [reply]
Hi. Sorry for long reply. Here is what in my CmakeList file : 
cmake_minimum_required(VERSION 3.31)
project(untitled)


set(CMAKE_CXX_STANDARD 20)

add_executable(untitled main.cpp)

**Reply by bret-walda** [reply]
so you know that cmake is a "build system" so you have to tell it what to do to your files to build them. you are missing everything that's why you couldn't build anything.
this is how i built it, it's not the most efficient use of it but this will get you started:
```
cmake_minimum_required(VERSION 3.14)
project(gui)
set(APP_SOURCES src/main.cpp)

set(EXTERNAL_SOURCES
libs/imgui.cpp
libs/imgui_demo.cpp
libs/imgui_draw.cpp
libs/imgui_tables.cpp
libs/imgui_widgets.cpp
libs/backends/imgui_impl_opengl3.cpp
libs/backends/imgui_impl_glfw.cpp
libs/glad/src/glad.c

)


add_executable(gui ${APP_SOURCES} ${EXTERNAL_SOURCES})

target_include_directories(gui PRIVATE
    libs
    libs/backends
    libs/include
    libs/GLFW
    libs/glad/include
)

target_link_libraries(gui PRIVATE 
G:/IDOWN/glfw-3.3.9/build/src/libglfw3.a
opengl32
)
```
1. Make sure to replace my paths with relative to where you have stored your imgui source and header files.
2. Link with glfw and opengl32 or whatever other API or backend you are using for example sdl2 or vulkan.
3.  You might need to compile gflw on your own but you don't strictly need it if you choose a different backend.

Hope this helps.

**Reply by romkor-lab-it** [reply]
Ohhh. Thank you so much! I will test it tomorrow and see if it will work.

**Reply by ori1202** [reply]
looking really promising, can you share a small git repo that shows how you create imgui window in main.cpp using this cmakelist file?

**Reply by bret-walda** [reply]
> looking really promising, can you share a small git repo that shows how you create imgui window in main.cpp using this cmakelist file?

I will post the main.cpp file in a minute can't create a repo right now.

**Reply by bret-walda** [reply]
This is the main.cpp file i messed around with the code a bit but didn't go any further.
```
#include "imgui.h"
#include "imgui_impl_glfw.h"
#include "imgui_impl_opengl3.h"
#include <stdio.h>
#define GL_SILENCE_DEPRECATION
#include <glad/glad.h>
#include <GLFW/glfw3.h>

bool window_maximized = false;

static void help_marker(const char* desc) {
    ImGui::TextDisabled("(?)");
    if (ImGui::IsItemHovered()) {
        ImGui::BeginTooltip();
        ImGui::PushTextWrapPos(ImGui::GetFontSize() * 35.0f);
        ImGui::TextUnformatted(desc);
        ImGui::PopTextWrapPos();
        ImGui::EndTooltip();
    }
}

static void glfw_error_callback(int error, const char* description) {
    fprintf(stderr, "GLFW Error %d: %s\n", error, description);
}

int main(int, char**) {
    // Set GLFW error callback
    glfwSetErrorCallback(glfw_error_callback);

    // Initialize GLFW
    if (!glfwInit()) {
        fprintf(stderr, "Failed to initialize GLFW\n");
        return 1;
    }

    // Decide GL+GLSL versions
    const char* glsl_version = "#version 130";
    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

    // Create window with graphics context
    GLFWwindow* window = glfwCreateWindow(1280, 720, "Hello, world with Dear ImGui", nullptr, nullptr);
    if (window == nullptr) {
        fprintf(stderr, "Failed to create GLFW window\n");
        glfwTerminate();
        return 1;
    }
    glfwMakeContextCurrent(window);
    glfwSwapInterval(1); // Enable vsync

    // Initialize GLAD
    if (!gladLoadGLLoader((GLADloadproc)glfwGetProcAddress)) {
        fprintf(stderr, "Failed to initialize GLAD\n");
        glfwDestroyWindow(window);
        glfwTerminate();
        return 1;
    }

    // Setup Dear ImGui context
    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO(); (void)io;
    io.ConfigFlags |= ImGuiConfigFlags_NavEnableKeyboard;

    // Setup Dear ImGui style
    ImGui::StyleColorsDark();
    ImGuiStyle& style = ImGui::GetStyle();
    style.WindowRounding = 0.0f;

    // Setup Platform/Renderer backends
    ImGui_ImplGlfw_InitForOpenGL(window, true);
    ImGui_ImplOpenGL3_Init(glsl_version);
    bool show = true;
    bool checkbox = false;

    // Main loop
    while (!glfwWindowShouldClose(window)) {
        glfwPollEvents();

        // Start the Dear ImGui frame
        ImGui_ImplOpenGL3_NewFrame();
        ImGui_ImplGlfw_NewFrame();
        ImGui::NewFrame();

        // Show a simple "Hello, world!" window
        ImGui::Begin("Hello, world!"); // Create a window called "Hello, world!" and append into it.
        ImGui::Text("Hello, world!");
        if (ImGui::Checkbox("visibility 3rd window", &checkbox)) {
            show = true;
        } else if (!checkbox) {
            show = false;
        }
        help_marker("hello");
        ImGui::End();

        // Render additional windows if 'show' is true
        if (show) {
            ImGui::Begin("Hello", &show);
            ImGui::Text("Another One");
            if (ImGui::Button("close me")) {
                show = false;
                checkbox = false;
            }
            ImGui::End();
        }

        // Main Window
        ImGui::Begin("Main Window", &show, ImGuiWindowFlags_MenuBar);

        // Menu bar
        if (ImGui::BeginMenuBar()) {
            if (ImGui::BeginMenu("File")) {
                if (ImGui::MenuItem("New")) {
                    // Implement your "New" functionality here
                    printf("Creating new file...\n");
                }
                if (ImGui::MenuItem("Open")) {
                    // Implement your "Open" functionality here
                    printf("Opening new file...\n");
                }
                if (ImGui::MenuItem("Save")) {
                    // Implement your "Save" functionality here
                    printf("Saving file...\n");
                }
                ImGui::EndMenu();
            }
            ImGui::EndMenuBar();
        }

        ImGui::Text("Hello, world!");

        ImGui::End(); // End main window

        // Rendering
        ImGui::Render();
        int display_w, display_h;
        glfwGetFramebufferSize(window, &display_w, &display_h);
        glViewport(0, 0, display_w, display_h);
        glClearColor(0.45f, 0.35f, 0.25f, 1.0f);
        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
        ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());

        glfwSwapBuffers(window);
    }

    // Cleanup
    ImGui_ImplOpenGL3_Shutdown();
    ImGui_ImplGlfw_Shutdown();
    ImGui::DestroyContext();
    glfwDestroyWindow(window);
    glfwTerminate();

    return 0;
}
```
